### PR TITLE
Feature/dimitris martin airen version

### PR DESCRIPTION
**Pull request by airenzp**
This PR started as the contribution of our group to the development guidelines. It contains minor improvements in different sections like Testing, it updates our top languages, according to git, and adds some useful links. Includes also new sections like Continuous Integration and Delivery as well as Reproducibility. Since this is my last day I am sending it as a whole, so I can get your reviews faster.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -90,9 +92,10 @@ out" (using upper-case variable names is a common way to do this).
 Comments should explain *why* the code does what it does. *What* it
 does should ideally already be evident from the code itself. If the code
 is cryptic and can't easily be simplified, explanations might well be
-needed.
+needed. A good comment clarifies intent.
 
-A good comment clarifies intent.
+Keep in mind that everything a developer needs to know to get started working in
+a project and everything a customer needs to run the project should be in README.md
```
Comment (pontus) [inline README.md:98]: This may be picking nits, but I'm somewhat sticky for correctness at times, and this is certainly impossible :)

How about something like:

> Try to capture and document as much as possible of what's needed to get started working in a project. Try also to
> capture the requirements and reasoning that explain larger architectural decisions.
>
> The customer also needs to know how to run the project, so be sure to document that, preferably with practical
> examples showing (e.g. rather than stating "the output is a container you can build and run", include actual
> example commands that will build and start the projects).

?

File: README.md [hunk 2]
```diff
@@ -273,7 +278,103 @@ should mention how to instantiate those variables/files, etc.
...
+the order in which you want to test them, and the expected responses) by a script instead of a human. Software testing involves the execution of a 
+software component or system component to evaluate one or more properties of interes. In general, these properties indicate the extent to which the component
+ or system under test:
+
+* meets the requirements that guided its design and development,
+* responds correctly to all kinds of inputs,
+* performs its functions within an acceptable time,
+* is sufficiently usable,
+* can be installed and run in its intended environments, and
+* achieves the general result its stakeholders desire.
```
Comment (pontus) [inline README.md:302]: :+1:

**Review by kostas-kou — APPROVED** [review]
I do not have any comments to make about the changes

File: README.md [hunk 3]
```diff
@@ -273,7 +278,103 @@ should mention how to instantiate those variables/files, etc.
...
+The developer's changes are validated by creating a build and running automated tests against the build. 
+By doing so, you avoid the integration hell that usually happens when people wait for release day to merge
+ their changes into the release branch.
+
+Continuous integration puts a great emphasis on testing automation to check that the application is not broken 
+whenever new commits are integrated into the main branch.
+
+#### What you need (cost)
+Your team will need to write automated tests for each new feature, improvement or bug fix.
+You need a continuous integration server that can monitor the main repository and run the tests automatically for every new commits pushed.
```
Comment (pontus) [inline README.md:340]: s/server/service/

Possibly add something about the cost of time taken after push, and whatever to differentiate tests when a change is "ready" vs still worked on.

File: README.md [hunk 4]
```diff
@@ -273,7 +278,103 @@ should mention how to instantiate those variables/files, etc.
...
+Developers need to merge their changes as often as possible, at least once a day.
```
Comment (pontus) [inline README.md:341]: This seems unrealistic. How about something about encouraging developers to limit the size of changes to make them easier to review and merge and working to merge changes as soon as they're ready?

Comment (airenzp) [reply README.md:341]: yeah, I agree.

File: README.md [hunk 5]
```diff
@@ -273,7 +278,103 @@ should mention how to instantiate those variables/files, etc.
...
+
+### Continuous delivery
+Continuous delivery is an extension of continuous integration to make sure that you can release new changes to your 
+customers quickly in a sustainable way. This means that on top of having automated your testing, you also have automated
+ your release process and you can deploy your application at any point of time by clicking on a button.
+ 
+#### What you need (cost)
+You need a strong foundation in continuous integration and your test suite needs to cover enough of your codebase.
+Deployments need to be automated. The trigger is still manual but once a deployment is started there shouldn't be a need for human intervention.
+Your team will most likely need to embrace feature flags so that incomplete features do not affect customers in production.
```
Comment (pontus) [inline README.md:355]: Belongs with continuous deployment?

File: README.md (hunk 3 above)
Comment (airenzp) [reply README.md:340]: Not sure what you mean here. Can you add more details?

File: README.md (hunk 5 above)
Comment (airenzp) [reply README.md:356]: I think we can actually remove continuous deployment and focus CI/CD. Removed now

File: README.md [hunk 6]
```diff
@@ -273,7 +278,103 @@ should mention how to instantiate those variables/files, etc.
...
+Building the release is easy as all integration issues have been solved early.
+Less context switching as developers are alerted as soon as they break the build and can work on fixing it before they move to another task.
+
+### Continuous delivery
+Continuous delivery is an extension of continuous integration to make sure that you can release new changes to your 
+customers quickly in a sustainable way. This means that on top of having automated your testing, you also have automated
+ your release process and you can deploy your application at any point of time by clicking on a button.
+ 
+#### What you need (cost)
+You need a strong foundation in continuous integration and your test suite needs to cover enough of your codebase.
```
Comment (pontus) [inline README.md:354]: Something about how good it is to have something consuming delivered releases in a timely manner to catch issues in a real environment that testing doesn't?

(It feels like I want to link this to agile development as well with customer interaction, but I'm not sure where/how.)

File: README.md (hunk 3 above)
Comment (pontus) [reply README.md]: Even if more testing is better, having time consuming tests may be frustrating and disturb the flow.

One example now is where we have tests taking roughly 10-15 minutes to run. I want to see those run successfully before I push, so after every time (I think) I'm done, there's a 15 minutes waiting period. As its long enough to not wait for, I need to remember to check on the tests and commit/push if everything's ready (if it's not, there will be another such cycle).

So, in some cases, it may be useful having a smaller set of test cases that are considered acceptable as a minimum threshold for what may be on a feature branch and a larger set of tests (typically labelled integration tests) that should pass before merge.

(And in some cases, that's not suitable, but it's worth noting the problem I think.)

Comment (airenzp) [reply README.md]: Mmm. Interesting, but I am not sure something is worth to mention in the main text. I do not know how to integrate it here at least. If you have a paragraph that works please send it, or some useful links.

**Review by inghylt — APPROVED** [review]

File: README.md [hunk 7]
```diff
@@ -600,3 +709,9 @@ and move philosophical, academic or otherwise unrelated technical
 discussions to an alternate forum.
 * Seek to understand the perspective of the author.
 * Sign off the final review with a thumbs up or some other positive remark.
+
+## Reproducibility guidelines
+
+Reproducibility is an important aspect of scientific research. As a support organisation directly involved in research we have a responsibility
+ to make sure that our work is reproducible. Please refer to the NBIS [Reproducibility guidelines](https://github.com/NBISweden/Reproducibility-Guidelines) 
+ for more information about this matter.
```
Comment (inghylt) [inline README.md:717]: @jbygdell Do you think this section and the referenced Reproducibility guidelines covers your thoughts in #18 ?

File: README.md [hunk 8]
```diff
@@ -273,7 +282,107 @@ should mention how to instantiate those variables/files, etc.
...
+
+
+#### Test driven development
+Test-driven development (TDD) is a software development process that relies on the repetition of a very short development cycle: requirements are turned into 
+very specific test cases, then the code is improved so that the tests pass. In simple terms, 
+test cases for each functionality are created and tested first and if the test fails then the new code is written in order to pass the test 
+and making code simple and bug-free.
+
+
+Useful links for writing tests in our most common  languages and frameworks:
```
Comment (danr) [inline README.md:328]: These links are not related to Test-driven development which this section is about.

File: README.md [hunk 9]
```diff
@@ -273,7 +282,107 @@ should mention how to instantiate those variables/files, etc.
...
+Integration testing tests integration between software modules. It is a software testing technique where individual units of a program 
+are combined and tested as a group. It checks the overall flow of the application after the integration of different modules.
+
+
+
+#### Test driven development
+Test-driven development (TDD) is a software development process that relies on the repetition of a very short development cycle: requirements are turned into 
+very specific test cases, then the code is improved so that the tests pass. In simple terms, 
+test cases for each functionality are created and tested first and if the test fails then the new code is written in order to pass the test 
+and making code simple and bug-free.
```
Comment (danr) [inline README.md:325]:
```suggestion
and to make the code simple and bug-free.
```
But introducing _more_ code is not typically associated with making code simpler or more bug-free.

File: README.md [hunk 10]
```diff
@@ -273,7 +282,107 @@ should mention how to instantiate those variables/files, etc.
...
+It is a testing method by which individual units of source code are tested to determine if they are ready to use. 
+It helps to reduce the cost of bug fixes since the bugs are identified during the early phases of the development lifecycle.
+
+What is Integration Test?
+Integration testing tests integration between software modules. It is a software testing technique where individual units of a program 
+are combined and tested as a group. It checks the overall flow of the application after the integration of different modules.
+
+
+
+#### Test driven development
```
Comment (danr) [inline README.md:321]:
```suggestion
#### Test-driven development
```

File: README.md [hunk 11]
```diff
@@ -273,7 +282,107 @@ should mention how to instantiate those variables/files, etc.
...
+
+#### What you gain
+Less bugs get shipped to production as regressions are captured early by the automated tests.
+Building the release is easy as all integration issues have been solved early.
+Less context switching as developers are alerted as soon as they break the build and can work on fixing it before they move to another task.
+
+### Continuous delivery
+Continuous delivery is an extension of continuous integration to make sure that you can release new changes to your 
+customers quickly in a sustainable way. This means that on top of having automated your testing, you also have automated
+ your release process and you can deploy your application at any point of time by clicking on a button.
```
Comment (danr) [inline README.md:363]: I suppose running a command instead of clicking a button goes as well? ;)

File: README.md [hunk 12]
```diff
@@ -273,7 +282,107 @@ should mention how to instantiate those variables/files, etc.
...
+
+#### What you need (cost)
+Your team will need to write automated tests for each new feature, improvement or bug fix. It should also  
+limit the size of the changes to make them easier to review and merge as soon as they're ready.
+You need a continuous integration server that can monitor the main repository and run the tests automatically. 
+
+#### What you gain
+Less bugs get shipped to production as regressions are captured early by the automated tests.
+Building the release is easy as all integration issues have been solved early.
+Less context switching as developers are alerted as soon as they break the build and can work on fixing it before they move to another task.
```
Comment (danr) [inline README.md:358]: I associate alerts with more context switching.

File: README.md [hunk 13]
```diff
@@ -273,7 +282,107 @@ should mention how to instantiate those variables/files, etc.
...
+
+Continuous integration puts a great emphasis on testing automation to check that the application is not broken 
+whenever new commits are integrated into the main branch.
+
+#### What you need (cost)
+Your team will need to write automated tests for each new feature, improvement or bug fix. It should also  
+limit the size of the changes to make them easier to review and merge as soon as they're ready.
+You need a continuous integration server that can monitor the main repository and run the tests automatically. 
+
+#### What you gain
```
Comment (danr) [inline README.md:355]: As a developer the main benefit for me is that there is one (reproducible) command (like `docker-compose up --build`) to get the entire project up and running locally for development.

File: README.md [hunk 14]
```diff
@@ -273,7 +282,107 @@ should mention how to instantiate those variables/files, etc.
...
+You need a strong foundation in continuous integration and your test suite needs to cover enough of your codebase.
+Deployments need to be automated. The trigger is still manual but once a deployment is started there shouldn't be a need for human intervention.
+Your team will most likely need to embrace feature flags so that incomplete features do not affect customers in production.
+#### What you gain
+The complexity of deploying software has been taken away. Your team doesn't have to spend days preparing for a release anymore.
+You can release more often, thus accelerating the feedback loop with your customers.
+There is much less pressure on decisions for small changes, hence encouraging iterating faster.
+
+
+### How the practices relate to each other
```
Comment (danr) [inline README.md:375]: I think this section is more confusing than helpful. I think this is the first time _continuous deployment_ is referenced. If you don't already know the concepts it's unlikely that the highlighted similaries and differences are helpful.

File: README.md [hunk 15]
```diff
@@ -273,7 +282,107 @@ should mention how to instantiate those variables/files, etc.
...
+To put it simply continuous integration is part of both continuous delivery and continuous deployment. 
+And continuous deployment is like continuous delivery, except that releases happen automatically.
+
+### Read CI/CD guides
```
Comment (danr) [inline README.md:379]: Maybe these guidies should be first. Then they can serve as an introduction as well as defining the terms for the rest of the text.

**Review by danr — COMMENTED** [review]
One thing that strikes me when reading this text is that concepts such as TDD and CI/CD are briefly introduced but it's unclear how these practices relate to NBIS development? Are these practices we implement? strive toward? are inspired by? ignore? The sections do not say.

File: README.md (hunk 8 above)
Comment (airenzp) [reply README.md:318]: I moved the CI/CD  section to Testing. In that way it does not have so much weight in the overall content of the dev guidelines. You ask me:
Are these practices we implement? strive toward? are inspired by? ignore? The sections do not say.
I guess we have to answer that together,  if they are not important enough for us maybe we can remove them from the guidelines. I personally think they are worth to be mentioned in the Testing section at least.

File: README.md (hunk 10 above)
Comment (airenzp) [reply README.md]: fixed

File: README.md (hunk 14 above)
Comment (airenzp) [reply README.md]: I agree. Removed this

File: README.md (hunk 9 above)
Comment (airenzp) [reply README.md]: fixed

File: README.md (hunk 11 above)
Comment (airenzp) [reply README.md]: improved

File: README.md (hunk 12 above)
Comment (airenzp) [reply README.md:355]: I used this practices in the past and found them useful. But this is just my personal experience.

Comment (danr) [reply README.md:355]: I also find them useful! That was not the point here. It was that this particular sentence seems to suggest that these alerts leads to less context switching. The important thing here is that this is a practice that might lead to that the build is broken less.

**Review by jonandernovella — APPROVED** [review]

**Review by danr — APPROVED** [review]

**Review by norling — APPROVED** [review]

**Review by jhagberg — APPROVED** [review]
